java-checks/src/main/java/org/sonar/java/checks/security/CryptographicKeySizeCheck.java

```diff
@@ -0,0 +1,151 @@
+          Integer minKeySize = findMinimumKeySizeFromAlgorithm(algorithm);
+          if (minKeySize != -1) {
+            checkAlgorithmParameterToReport(arg, minKeySize)
+              .ifPresent(reportString -> reportIssue(tree, reportString));
+          }
+        }
+      }
+    }
+
+    private boolean isKeyGenGetInstanceOrInitMethod(MethodInvocationTree mit) {
+      Symbol mitSymbol = mit.symbol();
+      if (mitSymbol.type() != null) {
+        String keyGenName = mitSymbol.type().fullyQualifiedName();
+        return ("javax.crypto.KeyGenerator".equals(keyGenName) || "java.security.KeyPairGenerator".equals(keyGenName))
+          && !mit.arguments().isEmpty();
+      }
+      return false;
+    }
+
+    private boolean isInitMethod(String methodName) {
+      return ("init".equals(methodName)) || ("initialize".equals(methodName));
+    }
+
+    private Optional<String> checkAlgorithmParameterToReport(ExpressionTree argument, Integer keySize) {
+      String reportString = null;
+      if (argument.is(Tree.Kind.INT_LITERAL) && (Integer.parseInt(((LiteralTree) argument).value()) < keySize)) {
+        reportString = "Use a key length of at least " + keySize + " bits.";
+      }
+      return Optional.ofNullable(reportString);
+    }
+
+    private int findMinimumKeySizeFromAlgorithm(String algorithm) {
+      switch (algorithm) {
+        case "Blowfish":
+          return 128;
+        case "RSA":
+          return 2048;
+        default:
+          return -1;
+      }
+    }
+
+    private Optional<Symbol> extractSymbolOfAssignedInstance(MethodInvocationTree mit) {
+      Tree parentExpression = mit.parent();
+      Symbol keyGenSymbol = null;
+      if (parentExpression.is(Tree.Kind.VARIABLE)) {
+        keyGenSymbol = ((VariableTree) parentExpression).symbol();
+      } else if (parentExpression.is(Tree.Kind.ASSIGNMENT)) {
+        ExpressionTree variable = ((AssignmentExpressionTree) parentExpression).variable();
+        if (variable.is(Tree.Kind.IDENTIFIER)) {
+          keyGenSymbol = ((IdentifierTree) variable).symbol();
+        } else if (variable.is(Tree.Kind.MEMBER_SELECT)) {
+          keyGenSymbol = ((MemberSelectExpressionTree) variable).identifier().symbol();
+        }
+      }
+      return Optional.ofNullable(keyGenSymbol);
+    }
+
+    private boolean initializeIsCalledOnPreviouslyDefinedKeyGen(MethodInvocationTree mit) {
+      boolean isValidKeyGenSymbol = false;
+      if (geInstanceKeyGenSymbol != null) {
+        ExpressionTree expr = ((MemberSelectExpressionTree) mit.methodSelect()).expression();
+        if (expr.is(Tree.Kind.IDENTIFIER)) {
+          isValidKeyGenSymbol = geInstanceKeyGenSymbol.equals(((IdentifierTree) expr).symbol());
+        } else if (expr.is(Tree.Kind.MEMBER_SELECT)) {
+          isValidKeyGenSymbol = geInstanceKeyGenSymbol.equals(((MemberSelectExpressionTree) expr).identifier().symbol());
```

Comment:
This logic duplicates a part of the previous method: in both cases you to get a `Symbol` from an `ExpressionTree`.